File: app/src/main/java/com/sony/open/cameratest/ZslReprocessActivity.java
/**
 * Copyright (c) 2018, Sony Mobile Communications Inc.
 * Licensed under the LICENSE.
 */
package com.sony.open.cameratest;

import android.Manifest;
import android.app.Activity;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.PackageManager;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.graphics.ImageFormat;
import android.graphics.SurfaceTexture;
import android.hardware.camera2.CameraCaptureSession;
import android.hardware.camera2.CameraCharacteristics;
import android.hardware.camera2.CameraDevice;
import android.hardware.camera2.CameraManager;
import android.hardware.camera2.CaptureFailure;
import android.hardware.camera2.CaptureRequest;
import android.hardware.camera2.TotalCaptureResult;
import android.hardware.camera2.params.InputConfiguration;
import android.hardware.camera2.params.StreamConfigurationMap;
import android.media.Image;
import android.media.ImageReader;
import android.media.ImageWriter;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.util.Log;
import android.util.Size;
import android.view.Surface;
import android.view.TextureView;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.ImageView;
import android.widget.Spinner;
import android.widget.Toast;

import java.nio.ByteBuffer;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.Locale;

@SuppressWarnings({"MissingPermission", "FieldCanBeLocal"})
public class ZslReprocessActivity extends Activity {
    private CameraManager        camManager;
    private CameraDevice         camDevice;
    private CameraCaptureSession camSession;
    private final static String camId = "0";

    private final int numUnprocessedImages = 50;
    private final int numProcessedImages   = 2;

    ImageReader irPreview, irReprocess;
    ImageWriter iwReprocess;

    // state
    private boolean isReady = false;
    List<Size> possibleSizes = new LinkedList<>();
    Size       captureSize;
    List<Surface> lSurfaces;

    private class ImagePair {
        TotalCaptureResult meta;
        Image              data;
    }

    // preview stream
    List<TotalCaptureResult> unprocessedMeta = new LinkedList<>();
    List<Image>              unprocessedData = new LinkedList<>();
    List<ImagePair> unprocessedImages = new ArrayList<>(3);

    // final results
    List<TotalCaptureResult> finalMeta = new LinkedList<>();
    List<Image>              finalData = new LinkedList<>();
    List<ImagePair> finalImages = new LinkedList<>();

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_zsl_reprocess);

        Button btnCapture = findViewById(R.id.btnZslReprocessCapture);
        btnCapture.setEnabled(false);

        // finish activity and notify user if camera permission missing
        if(checkSelfPermission(Manifest.permission.CAMERA) != PackageManager.PERMISSION_GRANTED) {
            Toast.makeText(this, "Camera permission not granted.", Toast.LENGTH_SHORT).show();
            finish();
        }

        // check camera and decide capture size
        try {
            camManager = (CameraManager) getSystemService(Context.CAMERA_SERVICE);
            CameraCharacteristics cc = camManager.getCameraCharacteristics(camId);

            int caps[] = cc.get(CameraCharacteristics.REQUEST_AVAILABLE_CAPABILITIES);
            if (!contains(caps, CameraCharacteristics.REQUEST_AVAILABLE_CAPABILITIES_PRIVATE_REPROCESSING)) {
                throw new Exception("PRIVATE_REPROCESSING not supported.");
            }

            StreamConfigurationMap map = cc.get(CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP);
            Size iSizes[] = map.getInputSizes(ImageFormat.PRIVATE);
            Size oSizes[] = map.getOutputSizes(ImageFormat.JPEG);
            for (Size o : oSizes) {
                for (Size i : iSizes) {
                    if (o.getHeight() == i.getHeight() && o.getWidth() == i.getWidth()) {
                        possibleSizes.add(o);
                    }
                }
            }
            int mw = 0, mh = 0;
            for (Size s : possibleSizes) {
                if (s.getHeight() > mh) {
                    mh = s.getHeight();
                    mw = s.getWidth();
                } else if (s.getHeight() == mh && s.getWidth() > mw) {
                    mw = s.getWidth();
                }
            }
            if (mw == 0 || mh == 0) {
                throw new Exception("no valid input/output size found.");
            }

            captureSize = new Size(mw, mh);
        } catch (Exception e) {
            Toast.makeText(this, "ERROR: Camera feature check failed:\n" + e.getMessage(), Toast.LENGTH_LONG).show();
            finish();
        }
    }

    @Override
    protected void onResume() {
        super.onResume();

        if(false) {
            // start preview now or when texture becomes ready (probably racy)
            TextureView.SurfaceTextureListener stl = new TextureView.SurfaceTextureListener() {
                @Override
                public void onSurfaceTextureAvailable(SurfaceTexture surfaceTexture, int i, int i1) {
                    startPreview();
                }

                @Override
                public void onSurfaceTextureSizeChanged(SurfaceTexture surfaceTexture, int i, int i1) {

                }

                @Override
                public boolean onSurfaceTextureDestroyed(SurfaceTexture surfaceTexture) {
                    return false;
                }

                @Override
                public void onSurfaceTextureUpdated(SurfaceTexture surfaceTexture) {

                }
            };
            TextureView tvPreview = findViewById(R.id.tvZslReprocessPreview);
            if (tvPreview.getSurfaceTexture() == null) {
                tvPreview.setSurfaceTextureListener(stl);
            } else {
                startPreview();
            }
        } else {
            // start preview after user has chosen resolution
            ArrayAdapter<String> adapter = new ArrayAdapter<>(this, android.R.layout.select_dialog_singlechoice);
            for(Size s : possibleSizes) {
                adapter.add(s.toString());
            }

            AlertDialog.Builder b = new AlertDialog.Builder(this);
            b.setTitle("Choose capture resolution");
            b.setAdapter(adapter, new DialogInterface.OnClickListener() {
                @Override
                public void onClick(DialogInterface dialogInterface, int i) {
                    captureSize = possibleSizes.get(i);
                    startPreview();
                }
            });
            b.setOnCancelListener(new DialogInterface.OnCancelListener() {
                @Override
                public void onCancel(DialogInterface dialogInterface) {
                    startPreview();
                }
            });
            b.show();
        }
    }

    @Override
    protected void onPause() {
        super.onPause();
        closeAll();
    }

    private void startPreview() {
        // destination surfaces
        lSurfaces = new LinkedList<>();
        TextureView tvPreview    = findViewById(R.id.tvZslReprocessPreview);
        SurfaceTexture stPreview = tvPreview.getSurfaceTexture();
        //stPreview.setDefaultBufferSize(captureSize.getWidth(), captureSize.getHeight());
        stPreview.setDefaultBufferSize(1280, 720);
        lSurfaces.add(new Surface(stPreview));
        irPreview = ImageReader.newInstance(captureSize.getWidth(), captureSize.getHeight(), ImageFormat.PRIVATE, numUnprocessedImages);
        irPreview.setOnImageAvailableListener(previewImageCallback, null);
        lSurfaces.add(irPreview.getSurface());
        irReprocess = ImageReader.newInstance(captureSize.getWidth(), captureSize.getHeight(), ImageFormat.JPEG, numProcessedImages);
        irReprocess.setOnImageAvailableListener(reprocessImageCallback, null);
        lSurfaces.add(irReprocess.getSurface());

        try {
            camManager.openCamera(camId, new CameraDevice.StateCallback() {
                @Override
                public void onOpened(@NonNull CameraDevice cameraDevice) {
                    camDevice = cameraDevice;
                    try {
                        InputConfiguration ic = new InputConfiguration(captureSize.getWidth(), captureSize.getHeight(), ImageFormat.PRIVATE);
                        cameraDevice.createReprocessableCaptureSession(ic, lSurfaces, new CameraCaptureSession.StateCallback() {
                            @Override
                            public void onConfigured(@NonNull CameraCaptureSession cameraCaptureSession) {
                                camSession = cameraCaptureSession;
                                iwReprocess = ImageWriter.newInstance(cameraCaptureSession.getInputSurface(), 2);

                                try {
                                    CaptureRequest.Builder b = camDevice.createCaptureRequest(CameraDevice.TEMPLATE_ZERO_SHUTTER_LAG);
                                    b.addTarget(lSurfaces.get(0));  // TextureView
                                    b.addTarget(lSurfaces.get(1));  // ImageReader irPreview
                                    cameraCaptureSession.setRepeatingRequest(b.build(), previewCaptureCallback, null);
                                } catch(Exception e) {
                                    Toast.makeText(ZslReprocessActivity.this, "ERROR: Failed to create preview stream:\n" + e.getMessage(), Toast.LENGTH_LONG).show();
                                    closeAll();
                                    finish();
                                }

                                Toast.makeText(ZslReprocessActivity.this,
                                        String.format(Locale.US, "Will capture at %dx%d.", captureSize.getWidth(), captureSize.getHeight()),
                                        Toast.LENGTH_SHORT).show();
                            }

                            @Override
                            public void onConfigureFailed(@NonNull CameraCaptureSession cameraCaptureSession) {
                                Toast.makeText(ZslReprocessActivity.this, "ERROR: Failed to configure session.", Toast.LENGTH_LONG).show();
                                closeAll();
                                finish();
                            }
                        }, null);

                    } catch(Exception e) {
                        Toast.makeText(ZslReprocessActivity.this, "ERROR: Failed to open session:\n" + e.getMessage(), Toast.LENGTH_LONG).show();
                        closeAll();
                        finish();
                    }
                }

                @Override
                public void onDisconnected(@NonNull CameraDevice cameraDevice) {
                    Toast.makeText(ZslReprocessActivity.this, "Camera was disconnected, closing.", Toast.LENGTH_SHORT).show();
                    closeAll();
                    finish();
                }

                @Override
                public void onError(@NonNull CameraDevice cameraDevice, int err) {
                    Toast.makeText(ZslReprocessActivity.this, "Camera received error " + err + ", closing." , Toast.LENGTH_SHORT).show();
                    closeAll();
                    finish();
                }
            }, null);
        } catch(Exception e) {
            Toast.makeText(this, "ERROR: Failed to open camera:\n" + e.getMessage(), Toast.LENGTH_LONG).show();
            finish();
        }
    }
    CameraCaptureSession.CaptureCallback previewCaptureCallback = new CameraCaptureSession.CaptureCallback() {
        long now, then = 0;
        double n = 0, s = 0, sq = 0;

        @Override
        public void onCaptureCompleted(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull TotalCaptureResult result) {
            super.onCaptureCompleted(session, request, result);

            if(then == 0) {
                // first frame
                then = System.nanoTime();
                return;
            }

            // gather statistics
            now = System.nanoTime();
            double diff = (now - then) / 1000000.0;
            n  += 1;
            s  += diff;
            sq += diff*diff;
            then = now;

            // print and reset regularly
            if(n >= 50) {
                // https://www-user.tu-chemnitz.de/~heha/hs/mr610.htm
                double avg = s / n;
                double sfq = sq - (s*s)/n;
                double std = Math.sqrt(sfq / (n - 1));
                Log.d("SRA", String.format(Locale.US, "preview at %.2f fps (%.2f ± %.2f ms)", 1000/avg, avg, std));

                n  = 0;
                s  = 0;
                sq = 0;
            }

            unprocessedMeta.add(result);
        }

        @Override
        public void onCaptureFailed(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull CaptureFailure failure) {
            super.onCaptureFailed(session, request, failure);
            Log.w("SRA", "preview failed");
        }

        @Override
        public void onCaptureBufferLost(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull Surface target, long frameNumber) {
            super.onCaptureBufferLost(session, request, target, frameNumber);
            Log.w("SRA", "preview lost buffer");
        }
    };
    ImageReader.OnImageAvailableListener previewImageCallback = new ImageReader.OnImageAvailableListener() {
        @Override
        public void onImageAvailable(ImageReader ir) {
            unprocessedData.add(ir.acquireNextImage());

            while(unprocessedMeta.size() >= 1 && unprocessedData.size() >= 1) {
                ImagePair nextImage = new ImagePair();
                nextImage.meta = unprocessedMeta.remove(0);
                nextImage.data = unprocessedData.remove(0);
                unprocessedImages.add(nextImage);

                if(unprocessedImages.size() > 1) {
                    ImagePair oldImage = unprocessedImages.remove(0);
                    oldImage.data.close();

                    if(!isReady) {
                        Button btn = findViewById(R.id.btnZslReprocessCapture);
                        btn.setEnabled(true);
                        isReady = true;
                    }
                }
            }
        }
    };

    public void btnCapture(View v) {
        if(!isReady || unprocessedImages.size() < 1) {
            Log.w("SRA", "Picture not taken: No unprocessed image available.");
            return;
        }

        try {
            ImagePair todo = unprocessedImages.remove(0);
            iwReprocess.queueInputImage(todo.data);
            CaptureRequest.Builder b = camDevice.createReprocessCaptureRequest(todo.meta);
            b.addTarget(lSurfaces.get(2));
            camSession.capture(b.build(), reprocessCaptureCallback, null);
        } catch(Exception e) {
            Toast.makeText(ZslReprocessActivity.this, "ERROR: Failed to reprocess:\n" + e.getMessage(), Toast.LENGTH_LONG).show();
            closeAll();
            finish();
        }
    }

    CameraCaptureSession.CaptureCallback reprocessCaptureCallback = new CameraCaptureSession.CaptureCallback() {
        @Override
        public void onCaptureCompleted(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull TotalCaptureResult result) {
            super.onCaptureCompleted(session, request, result);
            finalMeta.add(result);
        }

        @Override
        public void onCaptureFailed(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull CaptureFailure failure) {
            super.onCaptureFailed(session, request, failure);
            Log.w("SRA", "reprocess failed");
        }

        @Override
        public void onCaptureBufferLost(@NonNull CameraCaptureSession session, @NonNull CaptureRequest request, @NonNull Surface target, long frameNumber) {
            super.onCaptureBufferLost(session, request, target, frameNumber);
            Log.w("SRA", "reprocess lost buffer");
        }
    };
    ImageReader.OnImageAvailableListener reprocessImageCallback = new ImageReader.OnImageAvailableListener() {
        @Override
        public void onImageAvailable(ImageReader ir) {
            finalData.add(ir.acquireNextImage());

            while(finalMeta.size() > 0 && finalData.size() > 0) {
                ImagePair finalImage = new ImagePair();
                finalImage.meta = finalMeta.remove(0);
                finalImage.data = finalData.remove(0);

                if(finalImage.data.getFormat() == ImageFormat.JPEG) {
                    // retrieve bytes and release buffer
                    ByteBuffer buf = finalImage.data.getPlanes()[0].getBuffer();
                    byte[] jpegBytes = new byte[buf.remaining()];
                    buf.get(jpegBytes);
                    finalImage.data.close();

                    // decode and display
                    Bitmap bmpImage = BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.length-1);
                    ImageView ivThumbnail = findViewById(R.id.ivZslReprocessThumbnail);
                    ivThumbnail.setImageBitmap(bmpImage);
                } else {
                    Log.e("SRA", "format " + finalImage.data.getFormat() + " not supported");
                    finalImage.data.close();
                }
            }
        }
    };

    private void closeAll() {
        Button b = findViewById(R.id.btnZslReprocessCapture);
        b.setEnabled(false);

        isReady = false;
        captureSize = null;

        if(lSurfaces != null) {
            lSurfaces.clear();
            lSurfaces = null;
        }

        if(irPreview != null) {
            irPreview.close();
            irPreview = null;
        }

        if(camDevice != null) {
            if(camSession != null) {
                camSession.close();
                camSession = null;
            }
            camDevice.close();
            camDevice = null;
        }
    }
    private boolean contains(Size[] arr, int width, int height) {
        for(Size elem : arr) {
            if(elem.getWidth() == width && elem.getHeight() == height) {
                return true;
            }
        }
        return false;
    }
    private boolean contains(int[] arr, int val) {
        for(int elem : arr) {
            if(elem == val) {
                return true;
            }
        }
        return false;
    }
}

File: app/src/main/java/com/sony/open/cameratest/ZslReprocessActivity.kt
/*
 * Copyright (c) 2019, Sony Mobile Communications Inc.
 * Licensed under the LICENSE.
 */
package com.sony.open.cameratest

import android.app.Activity
import android.content.Context
import android.graphics.BitmapFactory
import android.graphics.ImageFormat
import android.hardware.camera2.*
import android.hardware.camera2.params.InputConfiguration
import android.hardware.camera2.params.StreamConfigurationMap
import android.media.Image
import android.media.ImageReader
import android.media.ImageWriter
import android.os.Bundle
import android.util.Log
import android.util.Size
import android.view.Surface
import android.view.TextureView
import android.view.View
import android.widget.Button
import android.widget.ImageView
import android.widget.Toast
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*
import kotlin.collections.ArrayList

class ZslReprocessActivity : Activity() {
    private val TAG = "ZslReprocessActivity"

    private val numUnprocessedImages = 50
    private val numProcessedImages = 2

    private val cameraManager by lazy { getSystemService(Context.CAMERA_SERVICE) as CameraManager }
    private val cameraHelper by lazy { CameraHelper(cameraManager) }

    private var cameraId: String? = null
    private var captureSize: Size? = null
    private var camDevice: CameraDevice? = null
    private var camSession: CameraCaptureSession? = null
    private var iwReprocess: ImageWriter? = null

    private class ImagePair(
            val meta: TotalCaptureResult,
            val img: Image
    )

    private val unprocessedMeta = LinkedList<TotalCaptureResult>()
    private val unprocessedData = LinkedList<Image>()
    private val unprocessedImages = LinkedList<ImagePair>()
    private val lSurfaces = ArrayList<Surface>(3)

    private val finalMeta = LinkedList<TotalCaptureResult>()
    private val finalData = LinkedList<Image>()

    private val closeables = LinkedList<AutoCloseable>()

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_zsl_reprocess)

        // find all camera devices with PRIVATE_REPROCESSING support
        val ids = cameraManager.cameraIdList.filter {
            val cc = cameraManager.getCameraCharacteristics(it)
            val caps = cc[CameraCharacteristics.REQUEST_AVAILABLE_CAPABILITIES]

            caps != null && caps.contains(CameraMetadata.REQUEST_AVAILABLE_CAPABILITIES_PRIVATE_REPROCESSING)
        }
        if (ids.isEmpty()) {
            Toast.makeText(this@ZslReprocessActivity, "No cameraId with PRIVATE_REPROCESSING support found.", Toast.LENGTH_LONG).show()
            finish()
            return
        }

        GlobalScope.launch(Dispatchers.Main) {
            // select cameraId to use
            cameraId = ids[0]
            if (ids.size > 1) {
                val ret = cameraHelper.askSelection(this@ZslReprocessActivity, "Choose cameraId:", ids)
                if (ret != null) {
                    cameraId = ids[ret]
                } else {
                    Toast.makeText(this@ZslReprocessActivity, "No camera selected.", Toast.LENGTH_LONG).show()
                    finish()
                    return@launch
                }
            }

            val cc = cameraManager.getCameraCharacteristics(cameraId!!)
            val map = cc[CameraCharacteristics.SCALER_STREAM_CONFIGURATION_MAP] as StreamConfigurationMap

            val iSizes = map.getInputSizes(ImageFormat.PRIVATE)
            val oSizes = map.getOutputSizes(ImageFormat.JPEG)

            val possibleSizes = iSizes.intersect(oSizes.asIterable()).sortedWith(compareBy(Size::getHeight, Size::getWidth).reversed())

            for (s in possibleSizes)
                Log.i(TAG, "Supported size: $s")

            if (possibleSizes.isEmpty()) {
                Toast.makeText(this@ZslReprocessActivity, "No matching input and output size found!", Toast.LENGTH_LONG).show()
                finish()
                return@launch
            }

            captureSize = possibleSizes.first()

            if (possibleSizes.size > 1) {
                val ret = cameraHelper.askSelection(this@ZslReprocessActivity, "Choose capture resolution:", possibleSizes.map { it.toString() })
                if (ret != null) {
                    captureSize = possibleSizes[ret]
                } else {
                    Toast.makeText(this@ZslReprocessActivity, "No capture resolution selected.", Toast.LENGTH_LONG).show()
                    finish()
                    return@launch
                }
            }

            // TODO: Instead move all selection logic to onResume?
            startPreview()
        }
    }

    override fun onResume() {
        super.onResume()
        startPreview()
    }

    override fun onPause() {
        super.onPause()
        closeAll()
    }

    @Suppress("UNUSED_PARAMETER")
    fun btnCapture(v: View) {
        val iwReprocess = iwReprocess ?: return
        val camDevice = camDevice ?: return
        val camSession = camSession ?: return

        val todo = unprocessedImages.poll()
        if (todo == null) {
            Toast.makeText(this@ZslReprocessActivity, "Picture not taken: No unprocessed image available.", Toast.LENGTH_LONG).show()
            return
        }

        // Queue image for reprocessing on captureSession input:
        todo.img.use(iwReprocess::queueInputImage)
        val b = camDevice.createReprocessCaptureRequest(todo.meta)
        // Send result to irReprocess surface:
        b.addTarget(lSurfaces[2]) /* reprocess surface */
        camSession.capture(b.build(), object : CameraCaptureSession.CaptureCallback() {
            override fun onCaptureCompleted(session: CameraCaptureSession, request: CaptureRequest, result: TotalCaptureResult) {
                super.onCaptureCompleted(session, request, result)
                Log.d(TAG, "ReprocessCapture.onCaptureCompleted")
                finalMeta.add(result)
                processFinalImage()
            }

            override fun onCaptureFailed(session: CameraCaptureSession, request: CaptureRequest, failure: CaptureFailure) {
                super.onCaptureFailed(session, request, failure)
                Log.w(TAG, "reprocess failed")
            }

            override fun onCaptureBufferLost(session: CameraCaptureSession, request: CaptureRequest, target: Surface, frameNumber: Long) {
                super.onCaptureBufferLost(session, request, target, frameNumber)
                Log.w(TAG, "reprocess lost buffer")
            }
        }, null)
    }

    /* Pop unprocessedMeta/unprocessedData and combine into unprocessedImages */
    private fun mergeUnprocessed() {
        while (unprocessedData.isNotEmpty() && unprocessedMeta.isNotEmpty()) {
            val meta = unprocessedMeta.poll()
            val img = unprocessedData.poll()
            if (meta == null || img == null)
                continue
            val p = ImagePair(meta, img)
            unprocessedImages.add(p)

            // TODO: single global variable?
            while (unprocessedImages.size > 1) {
                val old = unprocessedImages.poll()
                old?.img?.close()
            }

            val btnCapture = findViewById<Button>(R.id.btnZslReprocessCapture)
            btnCapture.isEnabled = true
        }
    }

    private fun processFinalImage() {
        while (finalMeta.isNotEmpty() && finalData.isNotEmpty()) {
            val meta = finalMeta.poll()
            val img = finalData.poll()
            if (img == null || meta == null)
                continue

            img.use {
                if (img.format == ImageFormat.JPEG) {
                    val start = System.currentTimeMillis()
                    // retrieve bytes and release buffer
                    val buf = img.planes[0].buffer
                    val jpegBytes = ByteArray(buf.remaining())
                    buf.get(jpegBytes)

                    // decode and display
                    val bmpImage = BitmapFactory.decodeByteArray(jpegBytes, 0, jpegBytes.size)
                    val ivThumbnail = findViewById<ImageView>(R.id.ivZslReprocessThumbnail)
                    ivThumbnail.rotation = 90f
                    ivThumbnail.setImageBitmap(bmpImage)
                    Log.i(TAG, "Decoding JPEG took ${(System.currentTimeMillis() - start)}ms")
                } else {
                    Log.e(TAG, "format ${img.format} not supported")
                }
            }
        }
    }

    private fun startPreview() {
        val tvPreview = findViewById<TextureView>(R.id.tvZslReprocessPreview)

        // Create immutable local copy
        val captureSize = captureSize
        if (captureSize == null) {
            Log.w(TAG, "No capture size defined, can't start!")
            return
        }
        val cameraId = cameraId
        if (cameraId == null) {
            Log.w(TAG, "No cameraId set, can't start!")
            return
        }

        Log.i(TAG, "Capturing $captureSize on camera $cameraId")

        val stPreview = tvPreview.surfaceTexture
        stPreview.setDefaultBufferSize(captureSize.width, captureSize.height);
        // stPreview.setDefaultBufferSize(1280, 720)
        val tvSurface = Surface(stPreview)
        lSurfaces.add(tvSurface)
        val irPreview = ImageReader.newInstance(captureSize.width, captureSize.height, ImageFormat.PRIVATE, numUnprocessedImages)
        closeables.add(irPreview)
        irPreview.setOnImageAvailableListener({ ir ->
            unprocessedData.add(ir.acquireNextImage())
            mergeUnprocessed()
        }, null)
        lSurfaces.add(irPreview.surface)
        val irReprocess = ImageReader.newInstance(captureSize.width, captureSize.height, ImageFormat.JPEG, numProcessedImages)
        closeables.add(irReprocess)
        irReprocess.setOnImageAvailableListener({ ir ->
            finalData.add(ir.acquireNextImage())
            Log.d(TAG, "irReprocess.onImageAvailable")
            processFinalImage()
        }, null)
        lSurfaces.add(irReprocess.surface)

        GlobalScope.launch(Dispatchers.Main) {
            // open camera device
            val device = cameraHelper.openCamera(cameraId)
            if (device == null) {
                Toast.makeText(this@ZslReprocessActivity, "Failed to open camera $cameraId.", Toast.LENGTH_LONG).show()
                return@launch
            }
            camDevice = device

            val ic = InputConfiguration(captureSize.width, captureSize.height, ImageFormat.PRIVATE)
            device.createReprocessableCaptureSession(ic, lSurfaces, object : CameraCaptureSession.StateCallback() {
                override fun onConfigured(cameraCaptureSession: CameraCaptureSession) {
                    camSession = cameraCaptureSession
                    val camInput = cameraCaptureSession.inputSurface
                    if (camInput == null) {
                        Toast.makeText(this@ZslReprocessActivity, "Camera capture session has no input surface.", Toast.LENGTH_LONG).show()
                        return
                    }
                    iwReprocess = ImageWriter.newInstance(camInput, 2)

                    val b = device.createCaptureRequest(CameraDevice.TEMPLATE_ZERO_SHUTTER_LAG)
                    // Show preview in UI:
                    b.addTarget(tvSurface) // TextureView
                    // Send unprocessed image to reader to cache for future processing
                    b.addTarget(irPreview.surface)
                    cameraCaptureSession.setRepeatingRequest(b.build(), object : CameraCaptureSession.CaptureCallback() {
                        var then = 0L
                        var n = 0.0
                        var s = 0.0
                        var sq = 0.0

                        override fun onCaptureCompleted(session: CameraCaptureSession, request: CaptureRequest, result: TotalCaptureResult) {
                            super.onCaptureCompleted(session, request, result)

                            if (then == 0L) {
                                // first frame
                                then = System.nanoTime()
                                return
                            }

                            // gather statistics
                            val now = System.nanoTime()
                            val diff = (now - then) / 1000000.0
                            n += 1
                            s += diff
                            sq += diff * diff
                            then = now

                            // print and reset regularly
                            if (n >= 50) {
                                // https://www-user.tu-chemnitz.de/~heha/hs/mr610.htm
                                val avg = s / n
                                val sfq = sq - (s * s) / n
                                val std = Math.sqrt(sfq / (n - 1))
                                Log.d(TAG, String.format(Locale.US, "preview at %.2f fps (%.2f ± %.2f ms)", 1000 / avg, avg, std))

                                n = 0.0
                                s = 0.0
                                sq = 0.0
                            }

                            unprocessedMeta.add(result)
                            mergeUnprocessed()
                        }

                        override fun onCaptureFailed(session: CameraCaptureSession, request: CaptureRequest, failure: CaptureFailure) {
                            super.onCaptureFailed(session, request, failure)
                            Log.w(TAG, "preview failed")
                        }

                        override fun onCaptureBufferLost(session: CameraCaptureSession, request: CaptureRequest, target: Surface, frameNumber: Long) {
                            super.onCaptureBufferLost(session, request, target, frameNumber)
                            Log.w(TAG, "preview lost buffer")
                        }
                    }, null)
                }

                override fun onConfigureFailed(cameraCaptureSession: CameraCaptureSession) {
                    Toast.makeText(this@ZslReprocessActivity, "ERROR: Failed to configure session.", Toast.LENGTH_LONG).show()
                    closeAll()
                    finish()
                }
            }, null)
        }
    }

    private fun closeAll() {
        val b = findViewById<Button>(R.id.btnZslReprocessCapture)
        b.isEnabled = false

//        for (s in lSurfaces)
//            s.release()

        for (c in closeables)
            c.close()
        closeables.clear()

        iwReprocess?.close()
        iwReprocess = null

        camSession?.close()
        camSession = null
        camDevice?.close()
        camDevice = null
    }
}
